File: DD_c_plusplus/Examples/SetCoverInstance/SetCoverInstance.cpp
#include "SetCoverInstance.h"

SetCoverInstance::SetCoverInstance(const string& file_path) {
    read_file(file_path);
}

SetCoverInstance::SetCoverInstance(dynamic_bitset<>* initial_state,
                                   vector<pair<string, vector<int>>> variables,
                                   vector<vector<int>> matrix_of_weight,
                                   vector<int> right_side_of_restrictions)
    : initial_state(initial_state),
      variables(std::move(variables)),
      matrix_of_weight(std::move(matrix_of_weight)),
      right_side_of_restrictions(std::move(right_side_of_restrictions))
{}

void SetCoverInstance::read_file(const string &file_path) {
    ifstream input(file_path);
    if( !input.is_open() ) {
        cerr << "Error: could not open file " << file_path << endl;
        exit(1);
    }

    int variable_length;
    int restrictions_length;

    input >> variable_length;
    input >> restrictions_length;

    initial_state = new dynamic_bitset<>(restrictions_length);
    for (int i = 0; i < restrictions_length; ++i) {
        initial_state->set(i);
    }

    right_side_of_restrictions = vector<int>(restrictions_length, 1);


    for (int i = 1; i <= variable_length; ++i) {
        std::string variable_name = "x_" + std::to_string(i);
        variables.emplace_back(variable_name, vector<int>{0,1});
    }

    for (int i = 0; i < variable_length; i++) {
        int value;
        input >> value;
        objective_weights.push_back(value);
    }

    for (int i = 0; i < restrictions_length; i++) {
        vector<int> row;
        for (int j = 0; j < variable_length; j++) {
            int value;
            input >> value;
            row.push_back(value);
        }
        matrix_of_weight.push_back(row);
    }

}

File: DD_python/Examples/SetCoverInstance/SetCoverInstance.py
from bitarray import bitarray
class SetCoverStructure:
    def __init__(self, file_path: str) -> None:
        """
        SourceCode for Set Cover Problem
        """
        self.initial_state: 'bitarray' = None
        self.variables: list[tuple[str, list[int]]] = None
        self.objective_weights:list[int] = None
        self.matrix_of_weight: list[list[int]] = None
        self.right_side_of_restrictions:list[int] = None

        self.read_file(file_path)

    def read_file(self, file_path: str) -> None:
        try:
            with open(file_path, 'r') as input_file:
                variable_length: int = int(input_file.readline().strip())
                restrictions_length: int = int(input_file.readline().strip())

                self.initial_state = bitarray(restrictions_length)
                self.initial_state.setall(1)
                self.right_side_of_restrictions: list[int] = [1] * restrictions_length

                self.objective_weights: list[int] = list(map(int, input_file.readline().strip().split()))

                self.matrix_of_weight = [
                    list(map(int, input_file.readline().strip().split()))
                    for _ in range(restrictions_length)
                ]

                self.variables = [("x_" + str(i + 1), [0, 1]) for i in range(variable_length)]

        except IOError:
            print(f"Error: could not open file {file_path}")
            exit(1)
    
    def __str__(self):
        return f"SetCoverStructure; initial_state:{self.initial_state}, variables:{self.variables}, objective_weights:{self.objective_weights}, matrix_of_weight:{self.matrix_of_weight}, right_side_of_restriction:{self.right_side_of_restrictions})"